endless-sky/endless-sky
# Fix bug that keeps player's flagship from being destroyed after an "on enter" conversation `explode` (and split up Ship::Move)

**Bugfix:** This PR fixes #8831 wherein ships that were destroyed (`hull = -1`) were able to heal for one frame. If the player's flagship had more than 60/s hull repair, the flagship would still be visible after the player died in an "on enter" conversation. 

## Fix Details
I've split up the gargantuan spaghetti mess `Ship::Move` into separate functions. The new `Ship::Move` works exactly like the old one, except that ships being destroyed can't do much.

## Testing Done
1. Fly around destroying things.
2. Make sure hyperspacing, landing, departing, afterburners, and other similar things worked.
3. Put this in data/[explodetest.txt](https://github.com/endless-sky/endless-sky/files/11702332/explodetest.txt)
4. Use a ship with good hull repair, such as [Friction Diction~3024-04-03 Main Campaign Done.txt](https://github.com/endless-sky/endless-sky/files/11654291/Friction.Diction.3024-04-03.Main.Campaign.Done.txt)
5. Depart
6. Visit the spaceport on Luna.
7. Agree to explode, and then depart Luna.
9. Your ship should explode, and you should be dead.

## Save File
See steps 3 through 9 in Testing Done.
